### PR TITLE
fixes Kfctl clientspec #2691

### LINKED ISSUE #2691: kfctl config issue
## Background
Currently, a KFApp has two layers:
1. parent app, the coordinator
2. child apps, a platform (gcp, minikube) and a packageManager (ksonnet, kustomize)

and different actions: init, generate, apply, delete, ..

When we run a command, e.g. `kfctl generate --A=B ...`, coordinator calls GcpApp.Generate and KsApp.Generate with the flags (the `option` variable in the code). 
I think it's not working properly now:
- some flag (e.g. UseBasicAuth) is needed is both platform and packageManager, and both generate and apply. 
- Gcp.Spec and KsApp.Spec stores different fields.
- I think it's confusing that whether a field, basicAuth, should be in `option`, or gcp.spec, or ks.spec.
- ksonnet.Generate right now has no information on whether to use basicAuth or not.

## Proposal
One possible solution is to save everything in `coordinator.spec`. And for any action (init or generate or else), coordinate passes `coordinator.spec` to platform and packageManager.
`gcpApp.Generate(coordinator.Spec)`

In addition, we can dump coordinator.spec to app.yaml. Then for any action, coordinator will
1. read app.yaml in as the spec
2. update the spec with new flags
3. pass spec to GcpApp and ksApp

This way users can just modify the app.yaml, and that will take effect.

cc @kkasravi @jlewi @kunmingg

### PR DESCRIPTION
**Pull request by kkasravi**
fixes #2691

- [x] move all attributes into ClientSpec
- [x] unmarshal app.yaml into ClientSpec for KfApp.{Apply, Generate, Delete}
- [x] use this ClientSpec when instantiating each KfApp
- [x] keep options local to coordinator
- [x] options for `kfctl init` are set in NewKfApp
- [x] options for `kfctl {generate, apply, delete}` are set in LoadKfApp

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2714)

### CODE REVIEW

**Comment by jlewi** [comment]
@kkasravi I started to review this but I think your commit history is messed up and this is making the diff difficult to read
  * It looks like you have commits from other authors that have already been checked in
  * There are changes to js and kustomize files that seem unrelated to this PR.
  * There's a whole bunch of files in pkg/apis/apps that appear to be deleted but they aren't on master either so I'm not sure why they show up.

**Comment by kkasravi** [comment]
It looks like this branch had a commit from @swiftdiaries, we'll try and not mix commits on the same branch - it's a huge headache.

**Comment by jlewi** [comment]
@kkasravi I think you need to sync now that #2705 is merged

**Comment by jlewi** [comment]
Looks like you still have changes from @swiftdiaries.

I think what you want to do is do an interactive rebase and drop the commits that were already committed to master 

This looks like a good tutorial
https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

You'll also likely have to deal with some conflicts between files.

**Comment by lluunn** [comment]
Where do we write the kdef to app.yaml?

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 1]
```diff
@@ -218,21 +126,16 @@ func getSA(name string, nameSuffix string, project string) string {
...
 
-func (gcp *Gcp) writeConfigFile(options map[string]interface{}) error {
-	buf, bufErr := yaml.Marshal(gcp.GcpApp)
+func (gcp *Gcp) writeConfigFile() error {
+	buf, bufErr := yaml.Marshal(gcp)
 	if bufErr != nil {
 		return bufErr
 	}
-	cfgFilePath := filepath.Join(gcp.GcpApp.Spec.AppDir, kftypes.KfConfigFile)
+	cfgFilePath := filepath.Join(gcp.Spec.AppDir, kftypes.KfConfigFile)
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Any benefits that GCP and ksonnet are sharing ```app.yaml``` file?

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 2]
```diff
@@ -875,37 +765,34 @@ func (gcp *Gcp) Generate(resources kftypes.ResourceEnum, options map[string]inte
...
-	setNameVal(&nv, "minioPd", gcp.GcpApp.Name+"-storage-artifact-store", false)
-	createConfigErr := gcp.writeConfigFile(options)
+		nv = gcp.Spec.ComponentParams["iap-ingress"]
+		setNameVal(&nv, "ipName", gcp.Spec.IpName, true)
+		setNameVal(&nv, "hostname", gcp.Spec.Hostname, true)
+	}
+	nv = gcp.Spec.ComponentParams["pipeline"]
+	setNameVal(&nv, "mysqlPd", gcp.Name+"-storage-metadata-store", false)
+	setNameVal(&nv, "minioPd", gcp.Name+"-storage-artifact-store", false)
+	createConfigErr := gcp.writeConfigFile()
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Will we lose part of "app.yaml" data written during ksonnet.Init()?
As gcp.go read/write less fields compared with ksonnet.go

**Comment by kunmingg** [comment]
In future if we can (partially) merge Init and Generate,
we can skip write/read ```app.yaml``` file on disk.
```apply``` can directly use ```generate``` outputs.

**Comment by lluunn** [comment]
/lgtm
/approve
/hold
for @kunmingg 
is the test failure the same issue @gabrielwen is fixing?

**Comment by kunmingg** [comment]
@lluunn 
Yeah, @kkasravi you might need to rebase after master is good.

**Comment by kkasravi** [comment]
ok, let me know @kunmingg @lluunn when I should rebase.

**Comment by kunmingg** [comment]
@kkasravi 
https://github.com/kubeflow/kubeflow/pull/2729 merged, can you rebase?

**Comment by kkasravi** [comment]
yep, asap.

**Comment by kkasravi** [comment]
@kunmingg rebased. Checking to see if the build passes.

**Comment by kkasravi** [comment]
resolved additional conflicts. Waiting on build status.

**Comment by gabrielwen** [comment]
from the error message of failing test, I suggest you to change from `Mkdir` to `MkdirAll` to make kfctl idempotent.

**Comment by kkasravi** [comment]
@kunmingg build now passes

**Comment by kunmingg** [comment]
/lgtm
/approve
Thanks! @kkasravi

**Pull request merged**